Source/ASDisplayNode.mm

```diff
@@ -948,9 +948,29 @@ - (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
                      restrictedToSize:(ASLayoutElementSize)size
                  relativeToParentSize:(CGSize)parentSize
 {
+  // Use a pthread specific to mark when this method is called re-entrant on same thread.
+  // We only want one calculateLayout signpost interval per thread.
+  // This is fast enough to do it unconditionally.
+  static pthread_key_t calculateLayoutKey;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    pthread_key_create(&calculateLayoutKey, NULL);
+  });
+  BOOL isNested = pthread_getspecific(calculateLayoutKey) != NULL;
+  if (!isNested) {
+    pthread_setspecific(calculateLayoutKey, kCFBooleanTrue);
+    ASProfilingSignpostStart(ASSignpostCalculateLayout, self, 0);
+  }
+
   ASSizeRange styleAndParentSize = ASLayoutElementSizeResolve(self.style.size, parentSize);
   const ASSizeRange resolvedRange = ASSizeRangeIntersect(constrainedSize, styleAndParentSize);
-  return [self calculateLayoutThatFits:resolvedRange];
+  ASLayout *result = [self calculateLayoutThatFits:resolvedRange];
+
+  if (!isNested) {
+    pthread_setspecific(calculateLayoutKey, NULL);
+    ASProfilingSignpostEnd(ASSignpostCalculateLayout, self, 0, ASSignpostColorDefault);
```

Comment:
I wonder if a higher-level version of this with fewer arguments would be warranted; ColorDefault is often used and the colors could be applied elsewhere as a mapping to the signpost type. A higher-level macro could also automatically capture the self pointer, and the signpost "routing layer" could choose to use it or not.